File: Android/src/com/jmedved/android/security/cryptography/bimil/BimilItem.java
package com.jmedved.android.security.cryptography.bimil;

import java.io.ByteArrayOutputStream;
import java.util.ArrayList;
import java.util.List;

public class BimilItem {

    private final BimilDocument Document;

    BimilItem(BimilDocument document) {
        this.Document = document;
        this.setTitle("");
        this._records = new ArrayList<BimilRecord>();
    }


    private int _iconIndex;
    public int getIconIndex() {
    	return this._iconIndex;
    }
    public void setIconIndex(int value) {
    	this._iconIndex = value;
    }
    
    private String _title;
    public String getTitle() {
    	return this._title;
    }
    public void setTitle(String value) {
    	this._title = value;
    }

    private List<BimilRecord> _records;
    public List<BimilRecord> getRecords() {
    	return _records;
    }


    public BimilRecord AddRecord(String key, String value, BimilRecordFormat format) {
    	BimilRecord record = new BimilRecord(this.Document, key, value, format);
    	this._records.add(record);
    	return record;
    }

    public BimilRecord AddTextRecord(String key, String value) {
        return AddRecord(key, value, BimilRecordFormat.Text);
    }

    public BimilRecord AddMultilineTextRecord(String key, String value) {
        return AddRecord(key, value, BimilRecordFormat.MultilineText);
    }

    public BimilRecord AddUrlRecord(String key, String value) {
        return AddRecord(key, value, BimilRecordFormat.Url);
    }

    public BimilRecord AddPasswordRecord(String key, String value) {
        return AddRecord(key, value, BimilRecordFormat.Password);
    }


    byte[] GetBytes() {
    	try {
    		ByteArrayOutputStream buffer = new ByteArrayOutputStream();
    		byte[] titleBytes = this.getTitle().getBytes("UTF8");
    		buffer.write(getInt32Bytes(titleBytes.length));
    		buffer.write(titleBytes);
    		buffer.write(getInt32Bytes(this.getIconIndex()));
    		for (BimilRecord record : this.getRecords()) {
    			byte[] keyBytes = record.getKey().getBytes();
    			byte[] valueBytes = record.getValue().getBytes();
    			buffer.write(getInt32Bytes(keyBytes.length));    			
    			buffer.write(getInt32Bytes(valueBytes.length));
    			buffer.write(getInt32Bytes(record.getFormat().getValue()));
    			buffer.write(keyBytes);
    			buffer.write(valueBytes);
	        }
    		return buffer.toByteArray();
    	} catch (Exception ex) {
    		throw new RuntimeException("Cannot save item.", ex);
    	}
    }


    static BimilItem Parse(BimilDocument document, byte[] buffer, int offset, int count) {
        try {
        	if (count < 4) { throw new RuntimeException("Invalid buffer size."); }

        	BimilItem res = new BimilItem(document);

        	int nameLength = getInt32(buffer, offset);
        	res.setTitle(new String(buffer, offset + 4, nameLength, "UTF8"));

        	res.setIconIndex(getInt32(buffer, offset + 4 + nameLength));

        	int currOffset = offset + 4 + nameLength + 4;
        	while (currOffset < (offset + count)) {
        		int keyLength = getInt32(buffer, currOffset);
        		int valueLength = getInt32(buffer, currOffset + 4);
        		BimilRecordFormat type = BimilRecordFormat.valueOf(getInt32(buffer, currOffset + 8));
        		BimilValue key = BimilValue.parse(document, buffer, currOffset + 12, keyLength);
        		BimilValue value = BimilValue.parse(document, buffer, currOffset + 12 + keyLength, valueLength);
        		res.getRecords().add(new BimilRecord(document, key.getText(), value.getText(), type));
        		currOffset += 12 + keyLength + valueLength;
        	}
        	if (currOffset != (offset + count)) { throw new RuntimeException("Invalid buffer content."); }

        	return res;
        } catch (Exception ex) {
            throw new RuntimeException("Cannot parse item.", ex);
        }
    }

    


    private static final byte[] getInt32Bytes(int value) {
    	return new byte[] { (byte)(value >>> 24), (byte)(value >>> 16), (byte)(value >>> 8), (byte)value};
    }
    
    private static final int getInt32(byte[] buffer, int offset) {
        return (buffer[offset+0] << 24) + ((buffer[offset+1] & 0xFF) << 16) + ((buffer[offset+2] & 0xFF) << 8) + (buffer[offset+3] & 0xFF);
    }

}

File: Windows/Source/Bimil/Document/BimilItem.cs
using System;
using System.Collections.Generic;
using System.Text;

namespace Medo.Security.Cryptography.Bimil {

    /// <summary>
    /// Item with encrypted records.
    /// </summary>
    public class BimilItem {

        private readonly BimilDocument Document;

        internal BimilItem(BimilDocument document) {
            this.Document = document;
            this.Records = new List<BimilRecord>();
        }


        /// <summary>
        /// Gets/sets icon index.
        /// </summary>
        public int IconIndex { get; set; }

        /// <summary>
        /// Gets/sets name.
        /// </summary>
        public string Name {
            get { return this.NameRecord.Value.Text; }
        }

        /// <summary>
        /// Gets name record.
        /// </summary>
        public BimilRecord NameRecord {
            get { return GetSystemRecord("Name"); }
        }

        /// <summary>
        /// Gets category record.
        /// </summary>
        public BimilRecord CategoryRecord {
            get { return GetSystemRecord("Category"); }
        }


        private void SetSystemRecordValue(string key, string value) {
            foreach (var record in this.Records) {
                if (record.Format == BimilRecordFormat.System) {
                    if (string.Equals(key, record.Key.Text, StringComparison.Ordinal)) {
                        record.Value.Text = value;
                        return;
                    }
                }
            }
            {
                var record = new BimilRecord(this.Document, key, value, BimilRecordFormat.System);
                this.Records.Add(record);
            }
        }


        /// <summary>
        /// Gets list of records.
        /// </summary>
        public IList<BimilRecord> Records { get; private set; }

        public void ClearNonSystemRecords() {
            for (int i = this.Records.Count - 1; i >= 0; i--) {
                if (this.Records[i].Format != BimilRecordFormat.System) {
                    this.Records.RemoveAt(i);
                }
            }
        }

        /// <summary>
        /// Adds new text record.
        /// </summary>
        /// <param name="key">Key.</param>
        /// <param name="value">Value.</param>
        public BimilRecord AddRecord(string key, string value, BimilRecordFormat format) {
            var record = new BimilRecord(this.Document, key, value, format);
            this.Records.Add(record);
            return record;
        }

        /// <summary>
        /// Adds new text record.
        /// </summary>
        /// <param name="key">Key.</param>
        /// <param name="value">Value.</param>
        public BimilRecord AddTextRecord(string key, string value) {
            return AddRecord(key, value, BimilRecordFormat.Text);
        }

        /// <summary>
        /// Adds new multiline text record.
        /// </summary>
        /// <param name="key">Key.</param>
        /// <param name="value">Value.</param>
        public BimilRecord AddMultilineTextRecord(string key, string value) {
            return AddRecord(key, value, BimilRecordFormat.MultilineText);
        }

        /// <summary>
        /// Adds new fixed-font record.
        /// </summary>
        /// <param name="key">Key.</param>
        /// <param name="value">Value.</param>
        public BimilRecord AddMonospacedTextRecord(string key, string value) {
            return AddRecord(key, value, BimilRecordFormat.MonospacedText);
        }

        /// <summary>
        /// Adds new URL record.
        /// </summary>
        /// <param name="key">Key.</param>
        /// <param name="value">Value.</param>
        public BimilRecord AddUrlRecord(string key, string value) {
            return AddRecord(key, value, BimilRecordFormat.Url);
        }

        /// <summary>
        /// Adds new password record.
        /// </summary>
        /// <param name="key">Key.</param>
        /// <param name="value">Value.</param>
        public BimilRecord AddPasswordRecord(string key, string value) {
            return AddRecord(key, value, BimilRecordFormat.Password);
        }


        internal byte[] GetBytes() {
            var buffer = new List<byte>(2048);
            foreach (var record in this.Records) {
                buffer.AddRange(GetInt32Bytes(record.Key.Bytes.Length));
                buffer.AddRange(GetInt32Bytes(record.Value.Bytes.Length));
                buffer.AddRange(GetInt32Bytes((int)record.Format));
                buffer.AddRange(record.Key.Bytes);
                buffer.AddRange(record.Value.Bytes);
            }
            return buffer.ToArray();
        }


        internal static BimilItem Parse(BimilDocument document, byte[] buffer, int offset, int count) {
            if (count < 4) { throw new FormatException("Invalid buffer size."); }

            var res = new BimilItem(document);

            int currOffset = offset;
            while (currOffset < (offset + count)) {
                int keyLength = GetInt32(buffer, currOffset);
                int valueLength = GetInt32(buffer, currOffset + 4);
                BimilRecordFormat type = (BimilRecordFormat)GetInt32(buffer, currOffset + 8);
                var key = BimilValue.Parse(document, buffer, currOffset + 12, keyLength);
                var value = BimilValue.Parse(document, buffer, currOffset + 12 + keyLength, valueLength);
                res.Records.Add(new BimilRecord(document, key.Text, value.Text, type));
                currOffset += 12 + keyLength + valueLength;
            }
            if (currOffset != (offset + count)) { throw new FormatException("Invalid buffer content."); }

            return res;
        }



        #region Helpers

        private static byte[] GetInt32Bytes(int value) {
            var bytes = BitConverter.GetBytes(value);
            if (BitConverter.IsLittleEndian) {
                return new byte[] { bytes[3], bytes[2], bytes[1], bytes[0] };
            } else {
                return bytes;
            }
        }

        private static int GetInt32(byte[] buffer, int offset) {
            if (BitConverter.IsLittleEndian) {
                return BitConverter.ToInt32(new byte[] { buffer[offset + 3], buffer[offset + 2], buffer[offset + 1], buffer[offset] }, 0);
            } else {
                return BitConverter.ToInt32(buffer, offset);
            }
        }

        private BimilRecord GetSystemRecord(string key) {
            foreach (var record in this.Records) {
                if (record.Format == BimilRecordFormat.System) {
                    if (string.Equals(key, record.Key.Text, StringComparison.Ordinal)) {
                        return record;
                    }
                }
            }
            {
                var record = new BimilRecord(this.Document, key, "", BimilRecordFormat.System);
                this.Records.Add(record);
                return record;
            }
        }

        #endregion

    }
}



// FORMAT (A128)
//--------
//  4x Key[0]-Length
//  4x Value[0]-Length
//  4x Type[0]
//  ?x Key[0]  (Value*)
//  ?x Value[0]  (Value*)
//  4x Key[1]-Length
//  4x Value[1]-Length
//  4x Type[1]
//  ?x Key[1]  (Value*)
//  ?x Value[1]  (Value*)
//  4x Key[n]-Length
//  4x Value[n]-Length
//  4x Type[n]
//  ?x Key[n]  (Value*)
//  ?x Value[n]  (Value*)
//
//   * This data is not encrypted (it is not needed since document itself is encrypted).
